src/BigQuery/BigQueryClient.php

```diff
@@ -0,0 +1,355 @@
+<?php
+/**
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\BigQuery;
+
+use Google\Cloud\BigQuery\Connection\ConnectionInterface;
+use Google\Cloud\BigQuery\Connection\Rest;
+
+/**
+ * Google Cloud BigQuery client. Allows you to create, manage, share and query
+ * data. Find more information at
+ * [Google Cloud BigQuery Docs](https://cloud.google.com/bigquery/what-is-bigquery).
+ */
+class BigQueryClient
+{
+    use JobConfigurationTrait;
+
+    const SCOPE = 'https://www.googleapis.com/auth/bigquery';
+    const INSERT_SCOPE = 'https://www.googleapis.com/auth/bigquery.insertdata';
+
+    /**
+     * @var ConnectionInterface $connection Represents a connection to BigQuery.
+     */
+    protected $connection;
+
+    /**
+     * @var string The project ID created in the Google Developers Console.
+     */
+    private $projectId;
+
+    /**
+     * Create a BigQuery client.
+     *
+     * Example:
+     * ```
+     * use Google\Cloud\BigQuery\BigQueryClient;
+     *
+     * $bigQuery = new BigQueryClient([
+     *     'projectId' => 'myAwesomeProject'
+     * ]);
+     * ```
+     *
+     * @param array $config {
+     *     Configuration options.
+     *
+     *     @type string $projectId The project ID from the Google Developer's
```

Comment:
The type-hinting accessibility provided by the builder pattern is definitely a plus, but I feel like the pattern comes with a few caveats in this use case.

I feel like it would become more difficult to use the class. As a user I would need to scour the class' available methods and then make a determination as to what is intended to set the object's initial state.

Also, with an array the class' configuration can easily be stored externally (in a .yaml file or such) and shared between clients. With the builder pattern, there would need to be a wrapper to provide the same functionality.

For this type of use case, I definitely lean towards arrays with the hope that eventually we will be able to port them to [named params](https://wiki.php.net/rfc/named_params) which will help solve the problem of type hinting and IDE compatability.